sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala

```diff
@@ -83,7 +83,19 @@ case class EnsureRequirements(conf: SQLConf) extends Rule[SparkPlan] {
         numPartitionsSet.headOption
       }
 
-      val targetNumPartitions = requiredNumPartitions.getOrElse(childrenNumPartitions.max)
+      // Read bucketed tables always obeys numShufflePartitions because maxNumPostShufflePartitions
+      // is usually much larger than numShufflePartitions,
+      // which causes some bucket map join lose efficacy after enabling adaptive execution.
```

Comment:
The comment is hard to understand. How about
```
If there are non-shuffle children that satisfy the required distribution, we have some tradeoffs
when picking the expected number of shuffle partitions:
1. we should avoid shuffling these children
2. we should have a reasonable parallelism

Here we pick the max number of partitions among these non-shuffle children as the expected number of shuffle partitions. However, if it's smaller than `conf.numShufflePartitions`, we pick `conf.numShufflePartitions` as the expected number of shuffle partitions.
```

Reply:
Done.